File: solutions/129.sum-root-to-leaf-numbers.js
/*
 * @lc app=leetcode id=129 lang=javascript
 *
 * [129] Sum Root to Leaf Numbers
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function (root, isRoot = true) {
  if (!root) return [];
  const left = sumNumbers(root.left, false);
  const right = sumNumbers(root.right, false);
  const res = [
    ...left.map((n) => String(root.val) + n),
    ...right.map((n) => String(root.val) + n),
  ];
  if (!res.length) res.push(String(root.val));
  return isRoot ? res.reduce((a, c) => a + Number(c), 0) : res;
};
// @lc code=end

File: solutions/129.sum-root-to-leaf-numbers.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def sum_numbers(root)
  res = 0
  tmp = ''
  pre_order = lambda do |node|
    tmp << node.val.to_s
    res += tmp.to_i if node.left.nil? && node.right.nil?
    pre_order.call(node.left) if node.left
    pre_order.call(node.right) if node.right
    tmp.chop!
  end
  pre_order.call(root)
  res
end
